lib/global-styles.php

```diff
@@ -0,0 +1,126 @@
+<?php
+/**
+ * Global styles functions, filters and actions, etc.
+ *
+ * @package gutenberg
+ */
+
+/**
+ * Function that given a branch of the global styles object recursively generates
+ * an array defining all the css vars that the global styles object represents.
+ *
+ * @param array  $global_styles_branch Array repreeting a brach of the global styles object.
+ * @param string $prefix Prefix to append to each variable.
+ * @return array An array whose keys are css variable names and whose values are the css variables value.
+ */
+function gutenberg_get_css_vars( $global_styles_branch, $prefix = '' ) {
+	$result = array();
+	foreach ( $global_styles_branch as $key => $value ) {
+		$processed_key = str_replace( '/', '-', $key );
+		$new_key       = ( $prefix ? $prefix . '-' : '' ) . $processed_key;
+
+		if ( is_array( $value ) ) {
+			$result = array_merge(
+				$result,
+				gutenberg_get_css_vars( $value, $new_key )
+			);
+		} else {
+			$result[ $new_key ] = $value;
+		}
+	}
+	return $result;
+}
+
+/**
+ * Function responsible for enqueuing the style that define the global styles css variables.
+ */
+function gutenberg_enqueue_global_styles_assets() {
+	$default_global_styles_path = dirname( dirname( __FILE__ ) ) . '/default-global-styles.json';
+	$default_global_styles      = null;
+
+	if ( file_exists( $default_global_styles_path ) ) {
+		$default_global_styles = json_decode(
+			file_get_contents( $default_global_styles_path ),
+			true
+		);
+	}
+
+	$theme_json_path          = locate_template( 'theme.json' );
+	$theme_json_global_styles = null;
+	if ( $theme_json_path ) {
+		$theme_json_global_styles = json_decode(
+			file_get_contents( $theme_json_path ),
+			true
+		);
+	}
+
+	// To-do: Load user customizations from a CPT.
+	$css_vars = array();
+	foreach (
+		array(
+			$default_global_styles,
+			$theme_json_global_styles,
+		) as $global_styles_definition
+	) {
+		if ( ! $global_styles_definition ) {
+			continue;
+		}
+		if ( isset( $global_styles_definition['global'] ) ) {
+			$css_vars = array_merge(
+				$css_vars,
+				gutenberg_get_css_vars( $global_styles_definition['global'], '--wp-gs' )
+			);
+		}
+		if ( isset( $global_styles_definition['blocks'] ) ) {
+			$css_vars = array_merge(
+				$css_vars,
+				gutenberg_get_css_vars( $global_styles_definition['blocks'], '--wp-gs' )
+			);
+		}
+	}
+
+	if ( empty( $css_vars ) ) {
+		return;
+	}
+
+	$inline_style = ":root {\n";
+	foreach ( $css_vars as $var => $value ) {
+		$inline_style .= "\t" . $var . ': ' . $value . ";\n";
+	}
+	$inline_style .= '}';
+
```

Comment:
A filter here would come in really handy...
I'd like to work on this issue: https://github.com/WordPress/gutenberg/issues/19611#issuecomment-579210156 but without a filter it's just going to get messy :wink:

Reply:
We are still in early stages, but I agree filters here make sense. I will propose an experimental filter 👍

Reply:
Shouldn't we hold on adding filters until we're sure about the APIs...

Reply:
It'd be good to start making a list of which functions (in the non-code sense) make sense to extend.